### PR TITLE
JENKINS-30375 Adding sauce on demand support

### PR DESCRIPTION
**Pull request by imuchnik**
Tests added to test all properties of sauce config
fixing docs

### CODE REVIEW

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/SauceOnDemandContext.groovy [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+           sauceConnectPath,
+           seleniumHost,
+           seleniumPort,
+           httpsProtocol,
+           options
+
+    List<String> webDriverBrowsers=[]
+    List<String> appiumBrowsers=[]
+
+    void appiumBrowsers(String... appiumBrowsers) {
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/SauceOnDemandContext.groovy]: Add GroovyDoc for each public method.

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/SauceOnDemandContext.groovy [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    void sauceConnectPath(String sauceConnectPath) {
+        this.sauceConnectPath = sauceConnectPath
+    }
+
+    void nativeAppPackage(String nativeAppPackage) {
+        this.nativeAppPackage = nativeAppPackage
+    }
+
+    void verboseLogging(boolean verboseLogging) {
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/SauceOnDemandContext.groovy]: Add a default value of `true` for all boolean parameters, e.g. `void verboseLogging(boolean verboseLogging = true) { ... }`

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext.groovy [hunk 3]
```diff
@@ -569,4 +569,44 @@ class WrapperContext extends AbstractExtensibleContext {
...
+                    'string'(browserName)
+                }
+            }
+
+            useLatestVersion(context.useLatestVersion)
+            launchSauceConnectOnSlave(context.launchSauceConnectOnSlave)
+            httpsProtocol(context.httpsProtocol ?: '')
+            options(context.options ?: '')
+            verboseLogging(context.verboseLogging)
+            condition(class: 'org.jenkins_ci.plugins.run_condition.core.AlwaysRun', plugin: 'run-condition@1.0')
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext.groovy]: Do not add the `plugin` attribute, the DSL does not generate them.

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext.groovy [hunk 4]
```diff
@@ -569,4 +569,44 @@ class WrapperContext extends AbstractExtensibleContext {
         node.append(context.selector)
         wrapperNodes << node
     }
+    /**
+     * Add support for SauceOnDemand plugin configuration
+     *
+     * @since 1.39
+     */
+
+    @RequiresPlugin(id = 'sauce-ondemand', minimumVersion = '@1.140')
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext.groovy]: The `@` in the version string seems to be incorrect.

**Comment by daspilker** [comment]
Thanks for your pull request. Please address the inline comments.

**Comment by imuchnik** [comment]
@daspilker Comments addressed. Thanks.

**Comment by imuchnik** [comment]
@daspilker Any ETA on this?

**Comment by daspilker** [comment]
Please rework the GroovyDoc. Fix the formatting and add complete sentences.

**Comment by imuchnik** [comment]
@daspilker Improved the docs, cleaned out deprecated values and rebased. Let me know what you think.

**Comment by imuchnik** [comment]
@daspilker Any change, this can make the next release?

File: docs/Home.md [hunk 5]
```diff
@@ -43,6 +45,7 @@ Have a look at the [Jenkins Job DSL Gradle example](https://github.com/sheehan/j
  * Added support for the [SSH Plugin](https://wiki.jenkins-ci.org/display/JENKINS/SSH+plugin)
    ([JENKINS-30957](https://issues.jenkins-ci.org/browse/JENKINS-30957))
 * 1.39 (October 05 2015)
+>>>>>>> 2838648a1ae24b46d50f9e35a5af9d6f21eb543c
```
Comment (daspilker) [inline docs/Home.md]: remove any merge artifacts

File: docs/Home.md [hunk 6]
```diff
@@ -18,7 +18,9 @@ If you want to get fancy you'll want to read up on [[configure block|The Configu
 Have a look at the [Jenkins Job DSL Gradle example](https://github.com/sheehan/job-dsl-gradle-example) to see how to organize a SCM repository for Job DSL scripts.
 
 ## Release Notes
+
```
Comment (daspilker) [inline docs/Home.md]: don't add additional whitespace

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/SauceOnDemandContext.groovy [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    List<String> appiumBrowsers = []
+
+    /**
+     * Adds list of browsers configured to run  test with Appium.If a single browser is selected,
+     * then it will be used to populate a SELENIUM_VERSION, SELENIUM_BROWSER, SELENIUM_PLATFORM and SELENIUM_DRIVER
+     * variable.
+     * If multiple browsers are selected, then they will be included in a SAUCE_ONDEMAND_BROWSERS environment variable
+     * in JSON format.
+     * These values can be extracted by your unit tests to invoke parallel Selenium tests.
+     * @param appiumBrowsers
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/SauceOnDemandContext.groovy]: remove all @param tags

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext.groovy [hunk 8]
```diff
@@ -586,4 +586,42 @@ class WrapperContext extends AbstractExtensibleContext {
         node.append(context.selector)
         wrapperNodes << node
     }
+    /**
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext.groovy]: add a newline between methods

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext.groovy [hunk 9]
```diff
@@ -586,4 +586,42 @@ class WrapperContext extends AbstractExtensibleContext {
...
+     * Add support for SauceOnDemand plugin configuration
+     *
+     * @since 1.40
+     */
+
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContext.groovy]: do not  add a new line between GroovyDoc and method

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/SauceOnDemandContext.groovy [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+    /**
+     * Location of the Sauce Connect executable that you wish the Sauce plugin to launch.
+     * Leave blank to use built-in Sauce Connect binary
+     * @param sauceConnectPath
+     */
+    void sauceConnectPath(String sauceConnectPath) {
+        this.sauceConnectPath = sauceConnectPath
+    }
+    /**
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/wrapper/SauceOnDemandContext.groovy]: always add a newline between methods

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContextSpec.groovy [hunk 11]
```diff
@@ -1018,6 +1018,87 @@ class WrapperContextSpec extends Specification {
         1 * mockJobManagement.requirePlugin('nodejs')
     }
 
+    def 'call sauce on demand with defaults'() {
+        when:
+        context.sauceOnDemandConfig {
+            webDriverBrowsers('foo', 'bar')
+            appiumBrowsers('larry', 'curly', 'moe')
+        }
+        then:
```
Comment (daspilker) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContextSpec.groovy]: add a newline before the label, not behind

**Comment by imuchnik** [comment]
@daspilker Thank you for your comment. It has been addressed and corrected. Any chances it will be merged before the next release. Thanks in advance.

**Comment by daspilker** [comment]
I rebased your changes and merged them manually.

**Comment by imuchnik** [comment]
Thanks!

**Pull request closed without merging**